netlogo-gui/src/main/agent/InRadiusOrCone.scala

```diff
@@ -0,0 +1,466 @@
+// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo
+
+package org.nlogo.agent
+
+import org.nlogo.api.{ AgentException, Dump }
+import org.nlogo.core.AgentKind
+
+import java.lang.{ Long => JLong }
+import java.util.{ ArrayList => JArrayList, HashSet => JHashSet, Iterator => JIterator, List => JList, RandomAccess }
+
+import scala.collection.mutable.ArrayBuffer
+
+object InRadiusOrCone {
+  /*
+   * A note on optimizations in this file.
+   *
+   * The optimizations here are performed based on relative benchmarked times for the following
+   * operations. These benchmarks were performed on Java 8u121. If `in-radius` performance suffers,
+   * rebenchmark these and adjust relative times.
+   */
+  /** Relative Amount of time to run `getPatchAtOffset` */
+  val CostToGetPatchAtOffset = 6
+  /** Relative Amount of time to run compute distance of turtle via world.protractor.distance */
+  val CostToComputeDistance = 4
+  /** Relative Amount of time to insert a Long (agent id) into a HashSet */
+  val CostToInsertID = 2
+  /** Relative Amount of time to check whether a HashSet contains a Long (agent id) */
+  val CostOfContainsID = 1
+
+  trait Strategy {
+    def name: String
+    def applies: Boolean
+    def cost: Long
+    def findAgents: IndexedAgentSet
+  }
+
+  trait RadialAgent[A <: Agent] {
+    def agentKind: AgentKind
+    def x(a: A): Double
+    def y(a: A): Double
+    def globalSet(w: World): AgentSet
+    def foreachAgent(p: Patch, f: A => Unit): Unit
+    def estimatedAgentCount(patchCount: Int, world: World): Long
+  }
+
+  implicit object RadialAgentPatch extends RadialAgent[Patch] {
+    val agentKind = AgentKind.Patch
+    def x(patch: Patch): Double = patch.pxcor
+    def y(patch: Patch): Double = patch.pycor
+    def globalSet(world: World): AgentSet = world.patches
+    def foreachAgent(p: Patch, f: Patch => Unit): Unit = f(p)
+    def estimatedAgentCount(patchCount: Int, world: World): Long = patchCount
+  }
+
+  implicit object RadialAgentTurtle extends RadialAgent[Turtle] {
+    val agentKind = AgentKind.Turtle
+    def x(turtle: Turtle): Double = turtle.xcor
+    def y(turtle: Turtle): Double = turtle.ycor
+    def globalSet(world: World): AgentSet = world.turtles
+    def foreachAgent(p: Patch, f: Turtle => Unit): Unit = {
+      p.turtlesHere match {
+        // this optimization may seem minor, but actually results in ~ 5-10% speedup
+        // when finding any turtles in-radius
+        case arrayList: JArrayList[_] =>
+          var i = 0
+          while (i < arrayList.size) {
+            f(arrayList.get(i).asInstanceOf[Turtle])
+            i += 1
+          }
+        case turtles =>
```

Comment:
Just to be clear, this branch can never be taken because `turtlesHere` always returns a `JArrayList`; you're just branching because it's only declared to return `Iterable`, right?